### ISSUE TITLE
NET-1843 Fix S1939 FP: Issue is raised when a class derives from another class and declares interface implemention

### ISSUE THREAD

**Issue by mary-georgiou-sonarsource** [SonarSource/sonar-dotnet#6823]
### Description

An FP is raised when a class is implementing a class and an interface - when the interface is not yet implemented by the class.

### Repro steps
```csharp
  class Foo { }

  interface IBar
  {
      int Test();
  }

  class Bar : Foo, IBar { } // Error [CS0535]
  // Noncompliant@-1 {{'Foo' implements 'IBar' so 'IBar' can be removed from the inheritance list.}}
```

### Expected behavior

An issue should not be raised as `Foo` is not implementing `IBar`

### Known workarounds

Implementing the interface in `IBar` in `Foo` stops raising the issue.

### Related information

* C#/VB.NET 8.53

**Comment by bottle2** [comment]
I just upgraded to 7.2.0.76209, the issue is still going

**Comment by arketec** [comment]
I have just also experienced the exact same issue with System.Text.Json.JsonConverter on a custom serializer that also implements an interface:

```cs
public interface IEnumConverter
{
    protected static readonly Dictionary<string, MyCustomEnum> conversions =
        new()
        {
           // default implementaion
        };
}

 public class MyEnumListConverter
     : System.Text.Json.Serialization.JsonConverter<List<MyCustomEnum>>,
         IEnumConverter
 {
    public override List<MyCustomEnum> Read(
     ref Utf8JsonReader reader,
     Type typeToConvert,
     JsonSerializerOptions options
    ) {
      if (reader.TokenType != JsonTokenType.StartArray)
      {
          throw new System.Text.Json.JsonException();
       }
    
      List<ConstructionYearType> list = new List<ConstructionYearType>();
    
      reader.Read();
      while (reader.TokenType != JsonTokenType.EndArray)
      {
          var val = reader.GetString();
          if (
              val != default
              && IEnumConverter.conversions.ContainsKey(val) // here it references the default implementation of the interface
          )
              list.Add(IEnumConverter.conversions[val]);
      
          reader.Read();
      }
      return list;
    }
 }
```

currently suppressing with System.Diagnostics.CodeAnalysis.SuppressMessage

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1843](https://sonarsource.atlassian.net/browse/NET-1843)

[NET-1843]: https://sonarsource.atlassian.net/browse/NET-1843?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
